In [1]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns 

sns.set()
%matplotlib inline 

##### Import the Csv file 

In [2]:
adult_df = pd.read_csv('adult.data.csv',
                        na_values = '?')

In [3]:
adult_df

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country salary  
0              2174             0              40  United-States  <=50K  
1                 0             0              13  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40           Cuba  <=50K  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States  <=50K  
32557             0             0              40  United-States   >50K  
32558             0             0              40  United-States  <=50K  
32559             0             0              20  United-States  <=50K  
32560         15024             0              40  United-States   >50K  

[32561 rows x 15 columns]

#### Races Represented in the data 

In [4]:
adult_df.race.value_counts()

White                 27816
Black                  3124
Asian-Pac-Islander     1039
Amer-Indian-Eskimo      311
Other                   271
Name: race, dtype: int64

In [5]:
adult_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       30725 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      30718 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  31978 non-null  object
 14  salary          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [6]:
adult_df.isna().sum()
# a significant number of values are missing from this data. All of which are object data types.

age                  0
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     583
salary               0
dtype: int64

### Average age of men 

In [7]:
# General Data Statistics
adult_df.describe() 

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [8]:
# Male Stats(Average age = 39.4)
# adult_df[adult_df.sex == 'Male'].sex.mean() = 39.4

adult_df[(adult_df.sex == 'Male')].describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  21790.000000  2.179000e+04   21790.000000  21790.000000  21790.000000   
mean      39.433547  1.917714e+05      10.102891   1329.370078    100.213309   
std       13.370630  1.067407e+05       2.662630   8326.312095    429.763097   
min       17.000000  1.376900e+04       1.000000      0.000000      0.000000   
25%       29.000000  1.182008e+05       9.000000      0.000000      0.000000   
50%       38.000000  1.801400e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.411462e+05      13.000000      0.000000      0.000000   
max       90.000000  1.455435e+06      16.000000  99999.000000   3770.000000   

       hours-per-week  
count    21790.000000  
mean        42.428086  
std         12.119755  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         49.000000  
max         99.000000

In [9]:
# Female stats 
adult_df[(adult_df.sex == 'Female')].describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  10771.000000  1.077100e+04   10771.000000  10771.000000  10771.000000   
mean      36.858230  1.857463e+05      10.035744    568.410547     61.187633   
std       14.013697  1.029861e+05       2.379954   4924.262944    340.907518   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       25.000000  1.173630e+05       9.000000      0.000000      0.000000   
50%       35.000000  1.760770e+05      10.000000      0.000000      0.000000   
75%       46.000000  2.283315e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    10771.000000  
mean        36.410361  
std         11.811300  
min          1.000000  
25%         30.000000  
50%         40.000000  
75%         40.000000  
max         99.000000

### Adults with Bachelors

In [10]:
# People with a Bachelors degree  

adults_with_bachelors = adult_df[adult_df.education == 'Bachelors']
adults_with_bachelors

age         workclass  fnlwgt  education  education-num  \
0       39         State-gov   77516  Bachelors             13   
1       50  Self-emp-not-inc   83311  Bachelors             13   
4       28           Private  338409  Bachelors             13   
9       42           Private  159449  Bachelors             13   
11      30         State-gov  141297  Bachelors             13   
...    ...               ...     ...        ...            ...   
32530   35               NaN  320084  Bachelors             13   
32531   30               NaN   33811  Bachelors             13   
32533   54           Private  337992  Bachelors             13   
32536   34           Private  160216  Bachelors             13   
32538   38           Private  139180  Bachelors             13   

           marital-status       occupation   relationship                race  \
0           Never-married     Adm-clerical  Not-in-family               White   
1      Married-civ-spouse  Exec-managerial        Husband               White   
4      Married-civ-spouse   Prof-specialty           Wife               Black   
9      Married-civ-spouse  Exec-managerial        Husband               White   
11     Married-civ-spouse   Prof-specialty        Husband  Asian-Pac-Islander   
...                   ...              ...            ...                 ...   
32530  Married-civ-spouse              NaN           Wife               White   
32531       Never-married              NaN  Not-in-family  Asian-Pac-Islander   
32533  Married-civ-spouse  Exec-managerial        Husband  Asian-Pac-Islander   
32536       Never-married  Exec-managerial  Not-in-family               White   
32538            Divorced   Prof-specialty      Unmarried               Black   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
0        Male          2174             0              40  United-States   
1        Male             0             0              13  United-States   
4      Female             0             0              40           Cuba   
9        Male          5178             0              40  United-States   
11       Male             0             0              40          India   
...       ...           ...           ...             ...            ...   
32530  Female             0             0              55  United-States   
32531  Female             0             0              99  United-States   
32533    Male             0             0              50          Japan   
32536  Female             0             0              55  United-States   
32538  Female         15020             0              45  United-States   

      salary  
0      <=50K  
1      <=50K  
4      <=50K  
9       >50K  
11      >50K  
...      ...  
32530   >50K  
32531  <=50K  
32533   >50K  
32536   >50K  
32538   >50K  

[5355 rows x 15 columns]

### Percentage of adults with Bachelors

In [11]:
adult_df['education'].value_counts(normalize=True) * 100


HS-grad         32.250238
Some-college    22.391818
Bachelors       16.446055
Masters          5.291607
Assoc-voc        4.244341
11th             3.608612
Assoc-acdm       3.276926
10th             2.865391
7th-8th          1.983969
Prof-school      1.768987
9th              1.578576
12th             1.329812
Doctorate        1.268389
5th-6th          1.022696
1st-4th          0.515955
Preschool        0.156629
Name: education, dtype: float64

#### `16.4%` of the adults have a Bachelor's Degree.

## Analyzing Adults with advanced education 
 

##### First task is to analyze the data. For adults with advanced education, earning over 50K

In [12]:
advanced_ed_mask = (adult_df.education == 'Bachelors') | (adult_df.education == 'Masters') | (adult_df.education=='Doctorate')

In [13]:
adult_df[advanced_ed_mask] # adults with advanced education

age         workclass  fnlwgt  education  education-num  \
0       39         State-gov   77516  Bachelors             13   
1       50  Self-emp-not-inc   83311  Bachelors             13   
4       28           Private  338409  Bachelors             13   
5       37           Private  284582    Masters             14   
8       31           Private   45781    Masters             14   
...    ...               ...     ...        ...            ...   
32538   38           Private  139180  Bachelors             13   
32539   71               NaN  287372  Doctorate             16   
32544   31           Private  199655    Masters             14   
32553   32           Private  116138    Masters             14   
32554   53           Private  321865    Masters             14   

           marital-status       occupation   relationship                race  \
0           Never-married     Adm-clerical  Not-in-family               White   
1      Married-civ-spouse  Exec-managerial        Husband               White   
4      Married-civ-spouse   Prof-specialty           Wife               Black   
5      Married-civ-spouse  Exec-managerial           Wife               White   
8           Never-married   Prof-specialty  Not-in-family               White   
...                   ...              ...            ...                 ...   
32538            Divorced   Prof-specialty      Unmarried               Black   
32539  Married-civ-spouse              NaN        Husband               White   
32544            Divorced    Other-service  Not-in-family               Other   
32553       Never-married     Tech-support  Not-in-family  Asian-Pac-Islander   
32554  Married-civ-spouse  Exec-managerial        Husband               White   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
0        Male          2174             0              40  United-States   
1        Male             0             0              13  United-States   
4      Female             0             0              40           Cuba   
5      Female             0             0              40  United-States   
8      Female         14084             0              50  United-States   
...       ...           ...           ...             ...            ...   
32538  Female         15020             0              45  United-States   
32539    Male             0             0              10  United-States   
32544  Female             0             0              30  United-States   
32553    Male             0             0              11         Taiwan   
32554    Male             0             0              40  United-States   

      salary  
0      <=50K  
1      <=50K  
4      <=50K  
5      <=50K  
8       >50K  
...      ...  
32538   >50K  
32539   >50K  
32544  <=50K  
32553  <=50K  
32554   >50K  

[7491 rows x 15 columns]

In [14]:
adult_df[~(advanced_ed_mask)] # adults without advanced education

age         workclass  fnlwgt     education  education-num  \
2       38           Private  215646       HS-grad              9   
3       53           Private  234721          11th              7   
6       49           Private  160187           9th              5   
7       52  Self-emp-not-inc  209642       HS-grad              9   
10      37           Private  280464  Some-college             10   
...    ...               ...     ...           ...            ...   
32556   27           Private  257302    Assoc-acdm             12   
32557   40           Private  154374       HS-grad              9   
32558   58           Private  151910       HS-grad              9   
32559   22           Private  201490       HS-grad              9   
32560   52      Self-emp-inc  287927       HS-grad              9   

              marital-status         occupation   relationship   race     sex  \
2                   Divorced  Handlers-cleaners  Not-in-family  White    Male   
3         Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
6      Married-spouse-absent      Other-service  Not-in-family  Black  Female   
7         Married-civ-spouse    Exec-managerial        Husband  White    Male   
10        Married-civ-spouse    Exec-managerial        Husband  Black    Male   
...                      ...                ...            ...    ...     ...   
32556     Married-civ-spouse       Tech-support           Wife  White  Female   
32557     Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558                Widowed       Adm-clerical      Unmarried  White  Female   
32559          Never-married       Adm-clerical      Own-child  White    Male   
32560     Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country salary  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
6                 0             0              16        Jamaica  <=50K  
7                 0             0              45  United-States   >50K  
10                0             0              80  United-States   >50K  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States  <=50K  
32557             0             0              40  United-States   >50K  
32558             0             0              40  United-States  <=50K  
32559             0             0              20  United-States  <=50K  
32560         15024             0              40  United-States   >50K  

[25070 rows x 15 columns]

In [15]:
# percentage of adults with higher education, earning over 50K 
adult_df[advanced_ed_mask].salary.value_counts(normalize=True) * 100

<=50K    53.464157
>50K     46.535843
Name: salary, dtype: float64

#### `46.5%` of adults with advanced education, earn 50K+ monthly.

### Adults without advanced education.

In [16]:
# ~(advanced_ed_mask) stands for all adults without Bachelors, Masters or a Doctorate.
adult_df[~(advanced_ed_mask)].salary.value_counts(normalize=True) * 100

<=50K    82.62864
>50K     17.37136
Name: salary, dtype: float64

#### `17.4%` of adults without advanced education earn over 50K monthly.

### Salary Analysis

In [17]:
# We will be analyzing based on salary.
# Adults earning above 50K, they are considered rich. 
# Salaries below or equal to 50K are not in the rich-category

In [18]:
salary_grouped = adult_df.groupby('salary')
rich_adults = salary_grouped.get_group('>50K')

### Adults earning 50K+ salary

In [19]:
rich_adults

age         workclass  fnlwgt     education  education-num  \
7       52  Self-emp-not-inc  209642       HS-grad              9   
8       31           Private   45781       Masters             14   
9       42           Private  159449     Bachelors             13   
10      37           Private  280464  Some-college             10   
11      30         State-gov  141297     Bachelors             13   
...    ...               ...     ...           ...            ...   
32539   71               NaN  287372     Doctorate             16   
32545   39         Local-gov  111499    Assoc-acdm             12   
32554   53           Private  321865       Masters             14   
32557   40           Private  154374       HS-grad              9   
32560   52      Self-emp-inc  287927       HS-grad              9   

           marital-status         occupation   relationship  \
7      Married-civ-spouse    Exec-managerial        Husband   
8           Never-married     Prof-specialty  Not-in-family   
9      Married-civ-spouse    Exec-managerial        Husband   
10     Married-civ-spouse    Exec-managerial        Husband   
11     Married-civ-spouse     Prof-specialty        Husband   
...                   ...                ...            ...   
32539  Married-civ-spouse                NaN        Husband   
32545  Married-civ-spouse       Adm-clerical           Wife   
32554  Married-civ-spouse    Exec-managerial        Husband   
32557  Married-civ-spouse  Machine-op-inspct        Husband   
32560  Married-civ-spouse    Exec-managerial           Wife   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
7                   White    Male             0             0              45   
8                   White  Female         14084             0              50   
9                   White    Male          5178             0              40   
10                  Black    Male             0             0              80   
11     Asian-Pac-Islander    Male             0             0              40   
...                   ...     ...           ...           ...             ...   
32539               White    Male             0             0              10   
32545               White  Female             0             0              20   
32554               White    Male             0             0              40   
32557               White    Male             0             0              40   
32560               White  Female         15024             0              40   

      native-country salary  
7      United-States   >50K  
8      United-States   >50K  
9      United-States   >50K  
10     United-States   >50K  
11             India   >50K  
...              ...    ...  
32539  United-States   >50K  
32545  United-States   >50K  
32554  United-States   >50K  
32557  United-States   >50K  
32560  United-States   >50K  

[7841 rows x 15 columns]

In [20]:
rich_adults['education'].value_counts(normalize=True) * 100

Bachelors       28.325469
HS-grad         21.362071
Some-college    17.689070
Masters         12.230583
Prof-school      5.394720
Assoc-voc        4.604005
Doctorate        3.902563
Assoc-acdm       3.379671
10th             0.790715
11th             0.765209
7th-8th          0.510139
12th             0.420865
9th              0.344344
5th-6th          0.204056
1st-4th          0.076521
Name: education, dtype: float64

In [21]:
(rich_adults['education'].value_counts(normalize=True) * 100)[['Bachelors', 'Masters', 'Doctorate']].sum()

44.45861497258004

#### `44.5%` of adults earning 50K+ in salary, have `advanced education`.

In [22]:
100 - (rich_adults['education'].value_counts(normalize=True) * 100)[['Bachelors', 'Masters', 'Doctorate']].sum()

55.54138502741996

#### `55.5%` of adults earning 50K+ in salary, do not have `advanced education`.

### Work-Hours Analysis 

In [23]:
# the minimum number of hours a person works per week 
adult_df.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

#### People who work for `1 hour` per week but still earn over 50K

In [24]:
min_hour_workers = adult_df[adult_df['hours-per-week'] == 1]

In [25]:
min_hour_workers.shape

(20, 15)

In [26]:
min_hour_workers['salary'].value_counts(normalize=True) * 100

<=50K    90.0
>50K     10.0
Name: salary, dtype: float64

#### `10%` of adults who work for 1 hour weekly, earn over 50K.

## Analyzing Salary by Country.

In [27]:
country_df = adult_df[['native-country','salary']]

In [28]:
country_df.groupby(['native-country', 'salary']).size().head(50)

native-country      salary
Cambodia            <=50K      12
                    >50K        7
Canada              <=50K      82
                    >50K       39
China               <=50K      55
                    >50K       20
Columbia            <=50K      57
                    >50K        2
Cuba                <=50K      70
                    >50K       25
Dominican-Republic  <=50K      68
                    >50K        2
Ecuador             <=50K      24
                    >50K        4
El-Salvador         <=50K      97
                    >50K        9
England             <=50K      60
                    >50K       30
France              <=50K      17
                    >50K       12
Germany             <=50K      93
                    >50K       44
Greece              <=50K      21
                    >50K        8
Guatemala           <=50K      61
                    >50K        3
Haiti               <=50K      40
                    >50K        4
Holand-Netherlands  <

In [29]:
# above 50K earners only.
country_df[country_df.salary == '>50K'].value_counts()

native-country      salary
United-States       >50K      7171
Philippines         >50K        61
Germany             >50K        44
India               >50K        40
Canada              >50K        39
Mexico              >50K        33
England             >50K        30
Italy               >50K        25
Cuba                >50K        25
Japan               >50K        24
Taiwan              >50K        20
China               >50K        20
Iran                >50K        18
South               >50K        16
France              >50K        12
Puerto-Rico         >50K        12
Poland              >50K        12
Jamaica             >50K        10
El-Salvador         >50K         9
Greece              >50K         8
Cambodia            >50K         7
Yugoslavia          >50K         6
Hong                >50K         6
Ireland             >50K         5
Vietnam             >50K         5
Ecuador             >50K         4
Haiti               >50K         4
Portugal            >50K    

In [30]:
from pandas import ExcelWriter
from pandas import ExcelFile

with pd.ExcelWriter('country_salary.xls') as ft:
    country_df.value_counts().to_frame().to_excel(ft, sheet_name='salaries_in_countries', encoding='UTF-8')
    


c:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3357: FutureWarning: As the xlwt package is no longer maintained, the xlwt engine will be removed in a future version of pandas. This is the only engine in pandas that supports writing in the xls format. Install openpyxl and write to an xlsx file instead. You can set the option io.excel.xls.writer to 'xlwt' to silence this warning. While this option is deprecated and will also raise a warning, it can be globally set and the warning suppressed.
  if (await self.run_code(code, result,  async_=asy)):


In [31]:
# Looking at India 

In [32]:
workers = adult_df[adult_df.salary == '>50K'].set_index('native-country')

In [33]:
workers.loc[['India'], ['occupation', 'salary']].groupby('occupation').size()

occupation
Adm-clerical         1
Exec-managerial      8
Other-service        2
Prof-specialty      25
Sales                1
Tech-support         2
Transport-moving     1
dtype: int64

In [34]:
top_indian_occupation = indian_workers.loc[['India'], ['occupation', 'salary']].groupby('occupation', as_index=True).size().sort_values(ascending=False).index[0]

NameError: name 'indian_workers' is not defined

#### The highest number of 50K+ earners in `India` work in `Prof-specialty`

In [ ]:
top_indian_occupation

'Prof-specialty'